File: LeetCode/addTwoNumbers.cpp
class Solution {
public:
    int carry = 0;\
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1 == NULL && l2 == NULL){
            if (carry == 1) {
                return new ListNode(1);
            }
            return NULL;
        }
        if (l1 == NULL) {
            int result = l2 -> val + carry;
            ListNode* current = new ListNode(result % 10);
            carry = result / 10;
            current -> next = addTwoNumbers(NULL, l2 -> next); 
            return current;
        }
        if (l2 == NULL) {
            int result = l1 -> val + carry;
            ListNode* current = new ListNode(result % 10);
            carry = result / 10;
            current -> next = addTwoNumbers(l1 -> next, NULL); 
            return current;
        }
        int result = l1 -> val + l2 -> val + carry;
        ListNode* current = new ListNode(result % 10);
        carry = result / 10;
        current -> next = addTwoNumbers(l1 -> next, l2 -> next); 
        return current;
    }
};

void trimLeftTrailingSpaces(string &input) {
    input.erase(input.begin(), find_if(input.begin(), input.end(), [](int ch) {
        return !isspace(ch);
    }));
}

void trimRightTrailingSpaces(string &input) {
    input.erase(find_if(input.rbegin(), input.rend(), [](int ch) {
        return !isspace(ch);
    }).base(), input.end());
}

vector<int> stringToIntegerVector(string input) {
    vector<int> output;
    trimLeftTrailingSpaces(input);
    trimRightTrailingSpaces(input);
    input = input.substr(1, input.length() - 2);
    stringstream ss;
    ss.str(input);
    string item;
    char delim = ',';
    while (getline(ss, item, delim)) {
        output.push_back(stoi(item));
    }
    return output;
}

ListNode* stringToListNode(string input) {
    // Generate list from the input
    vector<int> list = stringToIntegerVector(input);

    // Now convert that list into linked list
    ListNode* dummyRoot = new ListNode(0);
    ListNode* ptr = dummyRoot;
    for(int item : list) {
        ptr->next = new ListNode(item);
        ptr = ptr->next;
    }
    ptr = dummyRoot->next;
    delete dummyRoot;
    return ptr;
}

string listNodeToString(ListNode* node) {
    if (node == nullptr) {
        return "[]";
    }

    string result;
    while (node) {
        result += to_string(node->val) + ", ";
        node = node->next;
    }
    return "[" + result.substr(0, result.length() - 2) + "]";
}

int main() {
    string line;
    while (getline(cin, line)) {
        ListNode* l1 = stringToListNode(line);
        getline(cin, line);
        ListNode* l2 = stringToListNode(line);
        
        ListNode* ret = Solution().addTwoNumbers(l1, l2);

        string out = listNodeToString(ret);
        cout << out << endl;
    }
    return 0;
}

File: LeetCode/Add Two Numbers.java
// Title: Add Two Numbers
// Runtime: 20 ms
// Memory: 29.6 MB

class Solution {
    int carry = 0;
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null){
            if (carry == 1) {
                return new ListNode(1);
            }
            return null;
        }
        if (l1 == null) {
            int result = l2.val + carry;
            ListNode current = new ListNode(result % 10);
            carry = result / 10;
            current.next = addTwoNumbers(null, l2.next); 
            return current;
        }
        if (l2 == null) {
            int result = l1.val + carry;
            ListNode current = new ListNode(result % 10);
            carry = result / 10;
            current.next = addTwoNumbers(l1.next, null); 
            return current;
        }
        int result = l1.val + l2.val + carry;
        ListNode current = new ListNode(result % 10);
        carry = result / 10;
        current.next = addTwoNumbers(l1.next, l2.next); 
        return current;
    }
}

